### ISSUE TITLE
Database related rake tasks fails

### ISSUE THREAD

**Issue by dimonzozo** [SchemaPlus/schema_monkey_rails#1]
When I'm doing:

```
rake db:drop db:create
```

I'm getting error:

``` ruby
SchemaPlus::Views::Middleware::Schema::Tables::Mysql: No stack SchemaMonkey::Middleware::Schema::Tables
```

I've looked through source code and found that you handle other rake tasks like this:

``` ruby
SchemaMonkey::Rake.insert('db:schema:dump', 'db:schema:load')
```

**Comment by ronen** [comment]
@dimonzozo can you give me more details for how to recreate this (maybe include the output of `rake --trace db:drop db:create`)?  I'm not getting that error.

schema_plus 1.x only inserts itself for those two cases, which is why schema_monkey_rails does the same.  Though it does seem like it'd make more sense to insert itself in `app:environment` instead and handle all tasks that depend on rails.  Still I'd like to understand what's going on before making changes.

(Also could you try changing that to `SchemaMonkey::Rake.insert('app:environment')` and see if it fixes the problem for you?)

Thanks.

**Comment by dimonzozo** [comment]
Hi, @ronen. Some Rake tasks in db namespace is not loading environment and including `schema_monkey:insert` to `app:environment` is not useful. `db:drop` and `db:create` depends on `db:load_config` task. And `SchemaMonkey::Rake.insert('db:load_config')` seems fixing this issue.

For a now to get it work i've just added two lines to my project's Rakefile.

``` ruby
::Rake.application.lookup('db:drop').enhance(['schema_monkey:insert'])
::Rake.application.lookup('db:create').enhance(['schema_monkey:insert'])
```

This is a trace without these two lines:

```
> rake --trace db:drop db:create
```

```
** Invoke db:drop (first_time)
** Invoke db:load_config (first_time)
** Execute db:load_config
** Execute db:drop
SchemaPlus::Views::Middleware::Schema::Tables::Mysql: No stack SchemaMonkey::Middleware::Schema::Tables
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/stack.rb:9:in `rescue in insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/stack.rb:4:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:28:in `block in insert_middleware'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:25:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:25:in `insert_middleware'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:11:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:31:in `block (2 levels) in to_proc'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `inject'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `block in to_proc'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/monkey.rb:25:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/monkey.rb:25:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey.rb:34:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/active_record.rb:12:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:163:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:35:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:20:in `new'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:20:in `mysql2_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:438:in `new_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:448:in `checkout_new_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:422:in `acquire_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:349:in `block in checkout'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:348:in `checkout'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:263:in `block in connection'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:262:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:567:in `retrieve_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_handling.rb:113:in `retrieve_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_handling.rb:87:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/mysql_database_tasks.rb:8:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/mysql_database_tasks.rb:42:in `drop'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:114:in `drop'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:128:in `block in drop_current'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:275:in `block in each_current_configuration'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:274:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:274:in `each_current_configuration'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:127:in `drop_current'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/railties/databases.rake:28:in `block (2 levels) in <top (required)>'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:240:in `call'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:240:in `block in execute'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:235:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:235:in `execute'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `block in top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:78:in `block in run'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
./bin/rake:4:in `<main>'
Couldn't drop db_test
** Invoke db:create (first_time)
** Invoke db:load_config
** Execute db:create
SchemaPlus::Views::Middleware::Schema::Tables::Mysql: No stack SchemaMonkey::Middleware::Schema::Tables
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/stack.rb:9:in `rescue in insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/stack.rb:4:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:28:in `block in insert_middleware'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:25:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:25:in `insert_middleware'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/client.rb:11:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:31:in `block (2 levels) in to_proc'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `inject'
/Users/user/.rvm/gems/ruby-2.1.5/gems/its-it-1.1.1/lib/its-it/it.rb:30:in `block in to_proc'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/monkey.rb:25:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/monkey.rb:25:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey.rb:34:in `insert'
/Users/user/.rvm/gems/ruby-2.1.5/gems/schema_monkey-2.0.1/lib/schema_monkey/active_record.rb:12:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:163:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:35:in `initialize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:20:in `new'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:20:in `mysql2_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:438:in `new_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:448:in `checkout_new_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:422:in `acquire_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:349:in `block in checkout'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:348:in `checkout'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:263:in `block in connection'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:262:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:567:in `retrieve_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_handling.rb:113:in `retrieve_connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/connection_handling.rb:87:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/mysql_database_tasks.rb:8:in `connection'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/mysql_database_tasks.rb:16:in `create'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:93:in `create'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:107:in `block in create_current'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:275:in `block in each_current_configuration'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:274:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:274:in `each_current_configuration'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/tasks/database_tasks.rb:106:in `create_current'
/Users/user/.rvm/gems/ruby-2.1.5/bundler/gems/rails-62e9e61f2d1b/activerecord/lib/active_record/railties/databases.rake:17:in `block (2 levels) in <top (required)>'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:240:in `call'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:240:in `block in execute'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:235:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:235:in `execute'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
/Users/user/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `each'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:106:in `block in top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:78:in `block in run'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
/Users/user/.rvm/gems/ruby-2.1.5/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
./bin/rake:4:in `<main>'
```

**Comment by ronen** [comment]
@dimonzozo thanks for the details.  I must say I'm still confused as to what's going on...

I realized after posting my previous comment (and this is confirmed by the rake trace) that the error message comes from schema_monkey, which means that schema_monkey itself has been inserted somehow!

But the error message implies that schema_monkey didn't insert schema_plus_core before trying to insert schema_plus_views (since schema_plus_core would create the stack `SchemaMonkey::Middleware::Schema::Tables`).  That normally shouldn't be possible because schema_plus_views requires schema_plus_core which should guarantee the insertion order.

So I'm puzzled as to how this is happening.  How did schema_monkey get inserted?  But if it's inserted why isn't it loading the others properly?  And then why does explicitly adding the rake dependency fix it?  Hmm....

Is there any way you can generate a simple test case that reproduces the problem so I can debug it?  Or can you send me your application?

Thanks.

**Comment by dimonzozo** [comment]
Please try this steps to reproduce.

```
> rails -v
Rails 4.2.0
```

```
> rails new schema_monkey_test --database=mysql --skip-bundle --skip-spring --skip-test-unit
```

```
> cd schema_monkey_test 
> echo "gem 'schema_monkey_rails'\ngem 'schema_plus_views'" >> Gemfile
> bundle install && rake db:create
```

And I'm getting the error.

**Comment by ronen** [comment]
> Please try this steps to reproduce.

Thanks that does it.  (Strange, I tried an almost identical test case on my own and didn't get the error.)

Will look into what's happening...

**Comment by ronen** [comment]
Aha!  Actually this is good news :)

It turns out that schema_monkey was automatically inserting itself.... but just not quite doing it right.  I've fixed it so that it now automatically inserts itself properly.  Which means that schema_monkey_rails is no longer needed!

For a sanity check, can you try:
- use schema_monkey from master
- remove schema_monkey_rails from your Gemfile
- remove `::Rake.application.lookup('db:drop').enhance(['schema_monkey:insert'])` and `::Rake.application.lookup('db:create').enhance(['schema_monkey:insert'])` from your Rakefile

Once you've verified that it works for you too, i'll get rid of schema_monkey_rails....

**Comment by dimonzozo** [comment]
Everything works fine with `schema_monkey:master` and without `schema_monkey_rails`.

**Comment by ronen** [comment]
Great, thanks for verifying.

I've released schema_monkey 2.1.0 with this fix in it, and have retired schema_monkey_rails.

And thanks for spurring me to make this simplification!
